### PR TITLE
Investigate Jetpack's benchmarks library

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

tl;dr The library creates accurate benchmarking results as it is low level. It works well with command line and ci systems. I have been having trouble with development inside the benchmark module. It may be better to create a separate `benchmark-lib` for easy local development.

https://developer.android.com/jetpack/androidx/releases/benchmark
https://docs.google.com/document/d/1Xvli-XQNmZgo3phI5KPmj2tXxbLzNLiPCqUQ_Cd66GY/edit?usp=sharing

Android's jetpack has a benchmarking library. Benchmarked a couple minor operations. 
|  | do nothing | parse directions route | create simulated locations |
| -- | -- | -- | -- |
| minimum milliseconds | 0.000309 | 0.826 | 0.821 |
| maximum milliseconds | 0.000428 | 2.699 | 3.089 |
| median milliseconds | 0.000333 | 1.008 | 1.074 |

[com.mapbox.navigation.benchmark.test-benchmarkData.json.txt](https://github.com/mapbox/mapbox-navigation-android/files/6282006/com.mapbox.navigation.benchmark.test-benchmarkData.json.txt)

### Changelog

```
<changelog>Add a benchmarking module</changelog>
```

### Screenshots or Gifs

```
> Task :benchmark:connectedAndroidTest
04:06:32 I/Benchmark: Benchmark report files generated at /Users/kylemadsen/Development/mapbox/mapbox-navigation-android/benchmark/build/outputs/connected_android_test_additional_output

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.5.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 36s
412 actionable tasks: 6 executed, 406 up-to-date
```

### CODE REVIEW

**Comment by kmadsen** [comment]
cc: @tobrun @zugaldia @LukasPaczos

**Comment by zugaldia** [comment]
This looks very promising!

**Comment by tobrun** [comment]
Thank you for taking this for a spin, this library has been on my radar for a while but haven't been able to look more into. 
Would love to jump on a call and discuss future potential adoption of it.

**Comment by kmadsen** [comment]
Playing with places to launch the benchmarks. Doesn't make sense to launch it on firebase as it does not have a main launcher activity

```
➜  mapbox-navigation-android git:(km-add-benchmarks) ~/Downloads/google-cloud-sdk/bin/gcloud firebase test android run --app=benchmark/build/outputs/apk/androidTest/release/benchmark-release-androidTest.apk --device model=walleye,version=28,locale=de,orientation=landscape

Have questions, feedback, or issues? Get support by visiting:
  https://firebase.google.com/support/

Uploading [benchmark/build/outputs/apk/androidTest/release/benchmark-release-androidTest.apk] to Firebase Test Lab...
Raw results will be stored in your GCS bucket at [https://console.developers.google.com/storage/browser/test-lab-0m4kzyh5d96pw-yxcfsfu9yqry2/2021-05-18_14:58:47.863022_HmMc/]

Test [matrix-hlluvi772rwaa] has been created in the Google Cloud.
Firebase Test Lab will execute your robo test on 1 device(s).
Creating individual test executions...failed.
ERROR: (gcloud.firebase.test.android.run)
Matrix [matrix-hlluvi772rwaa] failed during validation: The app APK does not specify a main launcher activity.
```

**Comment by zugaldia** [comment]
Google I/O '21 introduced a new macrobenchmark library https://android-developers.googleblog.com/2021/05/whats-new-in-jetpack.html that is also worth looking into:

> The Macrobenchmark library extends Jetpack’s benchmarking coverage to app startup and integrated behaviors like scrolling performance. The library can be used remotely to track metrics in continuous integration testing or locally with profiling results viewable from Android Studio.
